app/lib/task/backend.dart

```diff
@@ -0,0 +1,881 @@
+      }
+
+      // List versions that not tracked, but should be
+      final untrackedVersions = versions
+          .map((v) => v.version!)
+          .whereNot((v) => state.versions!.containsKey(v))
+          .toList();
+
+      // List of versions that are tracked, but don't exist. These have
+      // probably been deleted.
+      final deletedVersions = state.versions!.keys
+          .where((v) => versions.none((pv) => pv.version == v))
+          .toList();
+
+      // There should never be an overlap between versions untracked and
+      // versions that tracked by now missing / deleted.
+      assert(
+        untrackedVersions.toSet().intersection(deletedVersions.toSet()).isEmpty,
+      );
+
+      // Stop transaction, if there is no changes to be made!
+      if (untrackedVersions.isEmpty && deletedVersions.isEmpty) {
+        return;
+      }
+
+      // Make changes!
+      state.versions!
+        // Remove versions that have been deleted
+        ..removeWhere((v, _) => deletedVersions.contains(v))
+        // Add versions we should be tracking
+        ..addAll({
+          for (final v in untrackedVersions)
+            v: PackageVersionState(
+              scheduled: DateTime(0),
+              attempts: 0,
+            ),
+        });
+
+      _log.info('Update state tracking for $packageName');
+      tx.insert(state);
+    });
+  }
+
+  /// Garbage collect [PackageState] and results from old runtimeVersions.
+  Future<void> garbageCollect() async {
+    // GC the old [PackageState] entities
+    await _db.deleteWithQuery(
+      _db.query<PackageState>()
+        ..filter('runtimeVersion <', gcBeforeRuntimeVersion),
+    );
+
+    // Limit to 50 concurrent deletion requests
+    final pool = Pool(50);
+
+    // Objects in the bucket are stored under the following pattern:
+    //   `<runtimeVersion>/<package>/<version>/...`
+    // Thus, we list with `/` as delimiter and get a list of runtimeVersions
+    await for (final d in _bucket.list(prefix: '', delimiter: '/')) {
+      if (!d.isDirectory) {
+        _log.warning('bucket should not contain any top-level object');
+        continue;
+      }
+
+      // Remove trailing slash from object prefix, to get a runtimeVersion
+      assert(d.name.endsWith('/'));
+      final rtVersion = d.name.substring(0, d.name.length - 1);
+
+      // Check if the runtimeVersion should be GC'ed
+      if (shouldGCVersion(rtVersion)) {
+        // List all objects under the `<rtVersion>/`
+        await for (final obj in _bucket.list(prefix: d.name, delimiter: '')) {
+          // Limit concurrency
+          final r = await pool.request();
+
+          // Schedule a microtask, that always ends by releasing the resource.
+          // Any issues deleting are logged as a warning, we'll probably try
+          // again later, so this is not really an issue.
+          scheduleMicrotask(() async {
+            try {
+              await _bucket.delete(obj.name);
```

Comment:
I don't think bucket listing will run into problems because we only list under the runtimeVersions we want to delete.

By not deleting with the `PackageState` we avoid any risk of forgetting to GC something because it wasn't referenced in the `PackageState` or vice versa.

If it's an old runtimeVersion then nobody uses it, and we should just delete it relying on as few invariants as possible. In this case I only rely on the prefix.

In any case, we have a few files per package version, so there won't be millions of files... like we had in the dartdoc bucket :D

Reply:
Also we would have to load the `dartdoc-index.json` file inorder to discover the `$id` in the name of `run-$id-dartdoc.blob` file.

And the code would be way more complicated :D